### PR TITLE
Rewrite npm publish

### LINKED ISSUE #606: Fail with an error if no platform build is found
**Steps**
```bash
npm i -g @tailcallhq/tailcall
```

Because all platform dependencies are optional, sometimes nothing gets installed. In that case, there should be an error issued stating that the platform is not yet supported, and the user should create an issue on our github repository.

### PR DESCRIPTION
**Pull request by neo773**
- Rewrote `gen` & `gen-root` in TypeScript
 - Changed naming convention for build packages to Node's

image

/claim #606
fixes #606

### CODE REVIEW

File: npm/generate-root.js [hunk 1]
```diff
@@ -62,18 +65,54 @@ async function genServerPackage(buildDefinitions) {
...
+    exec(\`npm install \${pkg}@\${optionalDependencies[pkg]} --no-save\`, (error, stdout, stderr) => {
+      if (error) {
+        console.error(\`Failed to install optional dependency: \${pkg}\`, stderr);
+      } else {
+        console.log(\`Successfully installed optional dependency: \${pkg}\`, stdout);
+      }
+    });
+  } else {
+    throw new Error(\`Unsupported platform \${platform}\ arch \${architecture}\`)
+  }
```
Comment (tusharmath) [inline npm/generate-root.js:115]: can we move this to a separate file and just copy it from generate-root.js

**Comment by amitksingh1490** [comment]
also generate.js need to generate correct os and cpu as of now its coming as `linux-gnu` `x86-x64` etc.

File: npm/generate-root.js (hunk 1 above)
Comment (neo773) [reply npm/generate-root.js]: @tusharmath Done

**Comment by amitksingh1490** [comment]
@neo773 
This doesn't work as expected
Steps to test
```bash
cd npm
npm install
npm run gen -- --target aarch64-apple-darwin --version 0.1.0 --build darwin-arm64
npm run gen-root -- --version 0.1.0
cd @tailcallhq/tailcall
npm link
```

Getting error
```/Users/amit/IdeaProjects/tailcall/npm/@tailcallhq/tailcall
   npm ERR! command failed
   npm ERR! command sh -c node ./scripts/installOptionalDeps.js
   npm ERR! node:internal/errors:490
   npm ERR!     ErrorCaptureStackTrace(err);
   npm ERR!     ^
   npm ERR!
   npm ERR! Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'detect-libc' imported from /Users/amit/IdeaProjects/tailcall/npm/@tailcallhq/tailcall/scripts/installOptionalDeps.js
   npm ERR!     at new NodeError (node:internal/errors:399:5)
   npm ERR!     at packageResolve (node:internal/modules/esm/resolve:889:9)
   npm ERR!     at moduleResolve (node:internal/modules/esm/resolve:938:20)
   npm ERR!     at defaultResolve (node:internal/modules/esm/resolve:1153:11)
   npm ERR!     at nextResolve (node:internal/modules/esm/loader:163:28)
   npm ERR!     at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
   npm ERR!     at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
   npm ERR!     at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
   npm ERR!     at link (node:internal/modules/esm/module_job:76:36) {
   npm ERR!   code: 'ERR_MODULE_NOT_FOUND'
   npm ERR! }
   npm ERR!
   npm ERR! Node.js v18.16.0

   npm ERR! A complete log of this run can be found in:
   npm ERR!     /Users/amit/.npm/_logs/2023-11-11T05_34_55_633Z-debug-0.log
   ```

**Comment by neo773** [comment]
@amitksingh1490 
Was trying out `preinstall` hook changed it back `postinstall`, should work now

**Comment by amitksingh1490** [comment]
> @amitksingh1490 Was trying out `preinstall` hook changed it back `postinstall`, should work now

```npm run gen-root -- --version 0.1.0

> gen-root
> tsx ./gen-root.ts --version 0.1.0

Generating package.json with version 0.1.0
node:internal/process/esm_loader:97
    internalBinding('errors').triggerUncaughtException(
                              ^

[Error: ENOENT: no such file or directory, open '/Users/amit/IdeaProjects/tailcall/npm/pre-install.js'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/Users/amit/IdeaProjects/tailcall/npm/pre-install.js'
}

Node.js v18.16.0
```getting error please try the steps in pervious comment and check if tc command works after that

**Comment by neo773** [comment]
@amitksingh1490 
Missed one path, sorry about that, could you try again now?

**Comment by amitksingh1490** [comment]
still getting this
```npm link
npm ERR! code 1
npm ERR! path /Users/amit/IdeaProjects/tailcall/npm/@tailcallhq/tailcall
npm ERR! command failed
npm ERR! command sh -c node ./scripts/installOptionalDeps.js
npm ERR! node:internal/errors:490
npm ERR!     ErrorCaptureStackTrace(err);
npm ERR!     ^
npm ERR! 
npm ERR! Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'detect-libc' imported from /Users/amit/IdeaProjects/tailcall/npm/@tailcallhq/tailcall/scripts/installOptionalDeps.js
npm ERR!     at new NodeError (node:internal/errors:399:5)
npm ERR!     at packageResolve (node:internal/modules/esm/resolve:889:9)
npm ERR!     at moduleResolve (node:internal/modules/esm/resolve:938:20)
npm ERR!     at defaultResolve (node:internal/modules/esm/resolve:1153:11)
npm ERR!     at nextResolve (node:internal/modules/esm/loader:163:28)
npm ERR!     at ESMLoader.resolve (node:internal/modules/esm/loader:838:30)
npm ERR!     at ESMLoader.getModuleJob (node:internal/modules/esm/loader:424:18)
npm ERR!     at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:77:40)
npm ERR!     at link (node:internal/modules/esm/module_job:76:36) {
npm ERR!   code: 'ERR_MODULE_NOT_FOUND'
npm ERR! }
npm ERR! 
npm ERR! Node.js v18.16.0

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/amit/.npm/_logs/2023-11-12T05_13_57_889Z-debug-0.log```
npm link should work

**Comment by neo773** [comment]
@amitksingh1490 
I'm not sure about `npm link` but i published a test package on npm registry and it's working as intended



https://github.com/tailcallhq/tailcall/assets/62795688/8cd18ef8-091a-499e-a43b-0cdd1f86011c

**Comment by neo773** [comment]
Separated preinstall hook from postinstall

Screenshot 2023-11-12 at 4 20 41 PM

**Review by amitksingh1490 — APPROVED** [review]

File: npm/post-install.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+const arch = process.arch
+
+let libcFamily
+family().then((fam) => {
+  libcFamily = fam
+})
+
+let libc
+if (platform === "win32") {
+  libc = "-msvc"
```
Comment (amitksingh1490) [inline npm/post-install.js:17]: Why is this needed?

Comment (neo773) [reply npm/post-install.js:17]: Well we have a "gnu" version of windows build so I hardcoded it for now since it also works for win32 arm64.
As for programmatically detecting it I couldn't find anything concrete and it seemed to involve writing native C code to use the relevant APIs.

**Review by amitksingh1490 — APPROVED** [review]
